File: internal/db/db_test.go
package db

import "testing"

func TestDBPutGetRestartRecovery(t *testing.T) {
	dir := t.TempDir()

	opts := DefaultOptions()
	opts.DataDir = dir
	opts.WALFileName = "wal.log"

	database, err := Open(opts)
	if err != nil {
		t.Fatalf("open: %v", err)
	}

	if err := database.Put([]byte("name"), []byte("mayank")); err != nil {
		t.Fatalf("put: %v", err)
	}
	if err := database.Close(); err != nil {
		t.Fatalf("close: %v", err)
	}

	reopened, err := Open(opts)
	if err != nil {
		t.Fatalf("reopen: %v", err)
	}
	defer reopened.Close()

	value, ok, err := reopened.Get([]byte("name"))
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if !ok {
		t.Fatal("expected recovered key")
	}
	if string(value) != "mayank" {
		t.Fatalf("unexpected value: %s", value)
	}
}

func TestDBFlushToSSTableAndReadBack(t *testing.T) {
	dir := t.TempDir()

	opts := DefaultOptions()
	opts.DataDir = dir
	opts.WALFileName = "wal.log"
	opts.MemtableThreshold = 2
	opts.SparseIndexGap = 1

	database, err := Open(opts)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	defer database.Close()

	if err := database.Put([]byte("a"), []byte("1")); err != nil {
		t.Fatalf("put a: %v", err)
	}
	if err := database.Put([]byte("b"), []byte("2")); err != nil {
		t.Fatalf("put b: %v", err)
	}
	if err := database.Put([]byte("c"), []byte("3")); err != nil {
		t.Fatalf("put c: %v", err)
	}

	value, ok, err := database.Get([]byte("a"))
	if err != nil {
		t.Fatalf("get a: %v", err)
	}
	if !ok || string(value) != "1" {
		t.Fatalf("unexpected value for a: %q ok=%v", value, ok)
	}

	value, ok, err = database.Get([]byte("c"))
	if err != nil {
		t.Fatalf("get c: %v", err)
	}
	if !ok || string(value) != "3" {
		t.Fatalf("unexpected value for c: %q ok=%v", value, ok)
	}
}

File: internal/db/db_test.cpp
#include "db.h"
#include <gtest/gtest.h>
#include <filesystem>
#include <thread>
#include <chrono>

using namespace bigdb::db;
namespace fs = std::filesystem;

class DBTest : public ::testing::Test {
protected:
    void SetUp() override {
        temp_dir = testing::TempDir() + "/db_test_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count());
        fs::create_directories(temp_dir);
    }
    void TearDown() override {
        fs::remove_all(temp_dir);
    }
    std::string temp_dir;
};

TEST_F(DBTest, PutGetRestartRecovery) {
    Options opts = DefaultOptions();
    opts.DataDir = temp_dir;
    opts.WALFileName = "wal.log";

    auto db = DB::Open(opts);
    ASSERT_NE(db, nullptr);

    EXPECT_NO_THROW(db->Put({'n','a','m','e'}, {'m','a','y','a','n','k'}));
    
    db->Close();

    auto reopened = DB::Open(opts);
    ASSERT_NE(reopened, nullptr);

    auto [value, ok] = reopened->Get({'n','a','m','e'});
    ASSERT_TRUE(ok) << "expected recovered key";
    
    std::string valStr(value.begin(), value.end());
    EXPECT_EQ(valStr, "mayank") << "unexpected value";

    reopened->Close();
}

TEST_F(DBTest, FlushToSSTableAndReadBack) {
    Options opts = DefaultOptions();
    opts.DataDir = temp_dir;
    opts.WALFileName = "wal.log";
    opts.MemtableThreshold = 2;
    opts.SparseIndexGap = 1;

    auto db = DB::Open(opts);
    ASSERT_NE(db, nullptr);

    EXPECT_NO_THROW(db->Put({'a'}, {'1'}));
    EXPECT_NO_THROW(db->Put({'b'}, {'2'}));
    EXPECT_NO_THROW(db->Put({'c'}, {'3'}));

    auto [valueA, okA] = db->Get({'a'});
    ASSERT_TRUE(okA);
    std::string valA(valueA.begin(), valueA.end());
    EXPECT_EQ(valA, "1");

    auto [valueC, okC] = db->Get({'c'});
    ASSERT_TRUE(okC);
    std::string valC(valueC.begin(), valueC.end());
    EXPECT_EQ(valC, "3");

    db->Close();
}
